backends/cuda-shared/ceed-cuda-shared-basis.c

```diff
@@ -0,0 +1,741 @@
+    grad3d(nelem, transpose, c_B, c_G, d_U, d_V, slice);
+  }
+}
+
+  /////////////
+ // Weights //
+/////////////
+__device__ void weight1d(const CeedInt nelem, const CeedScalar *qweight1d,
+                         CeedScalar *w) {
+  CeedScalar w1d[Q1D];
+  for (int i = 0; i < Q1D; ++i) {
+    w1d[i] = qweight1d[i];
+  }
+  for (int e = blockIdx.x * blockDim.x + threadIdx.x;
+       e < nelem;
+       e += blockDim.x * gridDim.x) {
+    for (int i = 0; i < Q1D; ++i) {
+      const int ind = e*Q1D + i;//sequential
+      w[ind] = w1d[i];
+    }
+  }
+}
+
+__device__ void weight2d(const CeedInt nelem, const CeedScalar *qweight1d,
+                         CeedScalar *w) {
+  CeedScalar w1d[Q1D];
+  for (int i = 0; i < Q1D; ++i) {
+    w1d[i] = qweight1d[i];
+  }
+  for (int e = blockIdx.x * blockDim.x + threadIdx.x;
+       e < nelem;
+       e += blockDim.x * gridDim.x) {
+    for (int i = 0; i < Q1D; ++i) {
+      for (int j = 0; j < Q1D; ++j) {
+        const int ind = e*Q1D*Q1D + i + j*Q1D;//sequential
+        w[ind] = w1d[i]*w1d[j];
+      }
+    }
+  }
+}
+
+__device__ void weight3d(const CeedInt nelem, const CeedScalar *qweight1d,
+                         CeedScalar *w) {
+  CeedScalar w1d[Q1D];
+  for (int i = 0; i < Q1D; ++i) {
+    w1d[i] = qweight1d[i];
+  }
+  for (int e = blockIdx.x * blockDim.x + threadIdx.x;
+       e < nelem;
+       e += blockDim.x * gridDim.x) {
+    for (int i = 0; i < Q1D; ++i) {
+      for (int j = 0; j < Q1D; ++j) {
+        for (int k = 0; k < Q1D; ++k) {
+          const int ind = e*Q1D*Q1D*Q1D + i + j*Q1D + k*Q1D*Q1D;//sequential
+          w[ind] = w1d[i]*w1d[j]*w1d[k];
+        }
+      }
+    }
+  }
+}
+
+extern "C" __global__ void weight(const CeedInt nelem,
+                                  const CeedScalar *__restrict__ qweight1d, CeedScalar *__restrict__ v) {
+  if (BASIS_DIM==1) {
+    weight1d(nelem, qweight1d, v);
+  } else if (BASIS_DIM==2) {
+    weight2d(nelem, qweight1d, v);
+  } else if (BASIS_DIM==3) {
+    weight3d(nelem, qweight1d, v);
+  }
+}
+
+);
+
+int CeedCudaInitInterp(CeedScalar *d_B, CeedInt P1d, CeedInt Q1d,
+                       CeedScalar **c_B);
+int CeedCudaInitInterpGrad(CeedScalar *d_B, CeedScalar *d_G, CeedInt P1d,
+                           CeedInt Q1d, CeedScalar **c_B_ptr, CeedScalar **c_G_ptr);
+
+int CeedBasisApplyTensor_Cuda_shared(CeedBasis basis, const CeedInt nelem,
+                               CeedTransposeMode tmode,
+                               CeedEvalMode emode, CeedVector u, CeedVector v) {
+  int ierr;
+  Ceed ceed;
+  ierr = CeedBasisGetCeed(basis, &ceed); CeedChk(ierr);
+  Ceed_Cuda_shared *ceed_Cuda;
+  CeedGetData(ceed, (void *) &ceed_Cuda); CeedChk(ierr);
+  CeedBasis_Cuda_shared *data;
+  CeedBasisGetData(basis, (void *)&data); CeedChk(ierr);
+  const CeedInt transpose = tmode == CEED_TRANSPOSE;
+  // const int optElems[7] = {0,32,8,3,2,1,8};
+  int elemsPerBlock = 1;//basis->Q1d < 7 ? optElems[basis->Q1d] : 1;
```

Comment:
future optimization work?

Reply:
Yes, but I need a proper benchmarking environment to decide which are the best.